docs/design-documents/hal/0002-pinmap-extension.md

```diff
@@ -0,0 +1,104 @@
+
+Any of the list mentioned above may be empty if none match the criteria.
+
+## Design choices
+
+These requirements do not impact the current API and thus they can all be implemented as an extension to the existing API.   
+It is proposed to introduce the following elements :
+- `DEVICE_PIN_EXTENDED_SUPPORT` A `device_has` flag indicating the implementation of this extension.
+- The following part of the API provides the features needed for requirements 3, 4 and 5
+  ```c
+  enum pinmap_form_factor_ {
+      PINMAP_FORM_FACTOR_ARDUINO_ZERO,
+      PINMAP_FORM_FACTOR_ARDUINO_DUE,
+      PINMAP_FORM_FACTOR_NXP_FRDM,
+      PINMAP_FORM_FACTOR_ST_MORPHO,
+      PINMAP_FORM_FACTOR_MTB,
+  } pinmap_form_factor_t;
+
+  /// returns a static NC terminated array of pins.
+  const PinName *pinmap_form_factor_pins(pinmap_form_factor_t form_factor);
+  /// returns a static NC terminated array of pins.
+  const PinName *pinmap_restricted_pins();
+  /// returns a null (\0) terminated static character array representing the name of the pin.
+  const char *pinmap_pin_to_string(PinName pin);
+  ```
+- These generic function make use of the existing API to answer requirements 1, 2 and 6.
+  ```c
+  /// Returns the `n`'th pin supporting the `peripheral` in the `map`.
+  PinName pinmap_find_peripheral_pin(const PinMap *map, int peripheral, uint32_t n);
+
+  /// Returns a pin from ((whitelist ∩ map) − blacklist).
+  /// Typically, whitelist will be the form factor list of pins.
+  /// The blacklist may be a list of pins already visited and/or tied to another peripheral.
+  PinName pinmap_find_peripheral_pin(const PinMap *map, const PinName *whitelist,
+                                      const PinName *blacklist, int per);
+
+  /// Verifies that `pin` is in `list`.
+  bool pinlist_has_pin(const PinName *list, PinName pin);
+
+  /// Applies pinmap_find_peripheral_pin to each map in maps ensuring a pin will not be used twice.
+  /// @param[in]  maps      A list of pinmap.
+  /// @param[in]  count     Number of elements of maps and pins.
+  /// @param[in]  whitelist An NC terminated list of pins.
+  /// @param[in]  blacklist An NC terminated list of pins.
+  /// @param[in]  per       A peripheral name.
+  /// @param[out] pins      An array of PinName where the result is stored.
+  bool pinmap_find_peripheral_pins(const PinMap const **maps,
+                                   uint32_t count,
+                                   const PinName *whitelist,
+                                   const PinName *blacklist,
+                                   uint32_t per,
+                                   PinName *pins);
+  ```
+- Additionally to these requirements any peripheral API must expose adequate functions to fetch an NC terminated list of pins associated to each of their their functions. These API extensions shall be gated with a `<PERIPHERAL_NAME>_PIN_EXTENTION` where `<PERIPHERAL_NAME>` is the name of the considered peripheral (`SPI`, `I2C`, `CAN` etc).
```

Comment:
Do we really need these specific wrappers? If they are not required by the solution why are we mentioning them? That not only adds unnecessary functions, but also pollutes "device_has" even more than it is already.

Reply:
Some device may require specific implementation of the wrapper.
Eg: K66F platform do not have a defined list of MOSI pins and MISO pins that work for both MASTER & SLAVE mode.
It has instead a SOUT and a SIN list of pins that are always respectively output or input.
a `spi_get_miso_pins(bool is_slave);` would select which pin list to return according to the is_slave argument.

A platform without this requirement would use the default implementation that returns SPI_PinMap_MISO.